vespalib/src/vespa/vespalib/util/ref_counted.h

```diff
@@ -0,0 +1,117 @@
+// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#pragma once
+
+#include <atomic>
+#include <concepts>
+#include <utility>
+
+// This file contains code that implements intrusive reference
+// counting with smart handles (ref_counted<T> points to T, T inherits
+// enable_ref_counted).
+
+// Functions with names starting with 'internal' are not intended for
+// direct use, but are available for completeness (enables incremental
+// re-write of code from bald to smart pointers).
+
+namespace vespalib {
+
+// This is the actual reference count. It cannot be moved or copied,
+// but it can be changed even when const. Classes that want to be
+// counted need to inherit from this. The destructor is virtual to
+// make sure the right one is called. This also enables the subref
+// function to directly delete the shared object.
+class enable_ref_counted
+{
+    static constexpr uint32_t MAGIC = 0xcc56a933;
+private:
+    uint32_t _guard;
+    mutable std::atomic<int32_t> _refs;
+protected:
+    enable_ref_counted() noexcept : _guard(MAGIC), _refs(1) {}
+public:
+    virtual ~enable_ref_counted() noexcept;
+    enable_ref_counted(enable_ref_counted &&) = delete;
+    enable_ref_counted(const enable_ref_counted &) = delete;
+    enable_ref_counted &operator=(enable_ref_counted &&) = delete;
+    enable_ref_counted &operator=(const enable_ref_counted &) = delete;
+    void internal_addref() const noexcept;
+    void internal_subref() const noexcept;
+    int32_t count_refs() const noexcept;
+};
+
+// This is the handle to a shared object. The handle itself is not
+// thread safe.
+template <typename T>
+class ref_counted
+{
+    // carefully placed here to give the best error messages
+    static_assert(std::derived_from<T,enable_ref_counted>);
+    template <typename X> friend class ref_counted;
+private:
+    T *_ptr;
+    ref_counted(T *ptr) noexcept : _ptr(ptr) {}
+    void maybe_subref() noexcept {
+        if (_ptr) {
+            _ptr->internal_subref();
+        }
+    }
+    static T *maybe_addref(T *ptr) noexcept {
+        if (ptr) {
```

Comment:
Would it be beneficial with `[[likely]]` on these branches in `maybe_subref`/`maybe_addref` (assuming the common case is that `_ptr` is present)?

Reply:
will do